File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/PokemonsFragment.java
package com.mdgd.pokemon.ui.pokemons;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.lifecycle.ViewModelProvider;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.mdgd.mvi.fragments.HostedFragment;
import com.mdgd.pokemon.PokemonsApp;
import com.mdgd.pokemon.R;
import com.mdgd.pokemon.models.filters.FilterData;
import com.mdgd.pokemon.models.repo.dao.schemas.PokemonFullDataSchema;

import java.util.Arrays;
import java.util.List;

import io.reactivex.rxjava3.disposables.CompositeDisposable;

public class PokemonsFragment extends HostedFragment<PokemonsContract.View, PokemonsContract.ViewModel, PokemonsContract.Host>
        implements PokemonsContract.View, View.OnClickListener, SwipeRefreshLayout.OnRefreshListener {

    private final CompositeDisposable onDestroyDisposables = new CompositeDisposable();
    private final PokemonsAdapter adapter = new PokemonsAdapter();
    private SwipeRefreshLayout refreshSwipe;
    private final EndlessScrollListener scrollListener = new EndlessScrollListener() {
        @Override
        public void onLoadMore(int page, int totalItemsCount, RecyclerView view) {
            if (refreshSwipe != null) {
                refreshSwipe.setRefreshing(true);
            }
            getModel().loadPage(page);
        }
    };
    private ImageButton filterAttack;
    private ImageButton filterDefence;
    private ImageButton filterSpeed;
    private View refresh;

    public static PokemonsFragment newInstance() {
        return new PokemonsFragment();
    }

    @Override
    public void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        onDestroyDisposables.add(adapter.getOnItemClickSubject().subscribe(pokemon -> getModel().onItemClicked(pokemon)));
    }

    @Override
    protected PokemonsContract.ViewModel createModel() {
        return new ViewModelProvider(this, new PokemonsViewModelFactory(PokemonsApp.getInstance().getAppComponent())).get(PokemonsViewModel.class);
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
        return inflater.inflate(R.layout.fragment_pokemons, container, false);
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
        super.onViewCreated(view, savedInstanceState);

        final RecyclerView recyclerView = view.findViewById(R.id.pokemons_recycler);
        recyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        recyclerView.addOnScrollListener(scrollListener);
        recyclerView.setAdapter(adapter);
        refresh = view.findViewById(R.id.pokemons_refresh);
        refreshSwipe = view.findViewById(R.id.pokemons_swipe_refresh);
        filterAttack = view.findViewById(R.id.pokemons_filter_attack);
        filterDefence = view.findViewById(R.id.pokemons_filter_defence);
        filterSpeed = view.findViewById(R.id.pokemons_filter_movement);

        refresh.setOnClickListener(this);
        refreshSwipe.setOnRefreshListener(this);

        filterAttack.setOnClickListener(this);
        filterDefence.setOnClickListener(this);
        filterSpeed.setOnClickListener(this);
    }

    @Override
    public void onDestroyView() {
        refresh = null;
        refreshSwipe = null;
        filterAttack = null;
        filterDefence = null;
        filterSpeed = null;
        super.onDestroyView();
    }

    @Override
    public void proceedToNextScreen(long id) {
        if (hasHost()) {
            getFragmentHost().proceedToPokemonScreen(id);
        }
    }

    @Override
    public void updateFilterButtons(boolean isActive, String filter) {
        final ImageButton view;
        switch (filter) {
            case FilterData.FILTER_ATTACK:
                view = filterAttack;
                break;

            case FilterData.FILTER_DEFENCE:
                view = filterDefence;
                break;

            case FilterData.FILTER_SPEED:
                view = filterSpeed;
                break;
            default:
                view = null;
        }
        if (view == null) {
            return;
        }
        if (isActive) {
            view.setColorFilter(ContextCompat.getColor(requireContext(), R.color.filter_active));
        } else {
            view.setColorFilter(ContextCompat.getColor(requireContext(), R.color.filter_inactive));
        }
    }

    @Override
    public void onClick(View view) {
        if (view == refresh) {
            if (!refreshSwipe.isRefreshing()) {
                refreshSwipe.setRefreshing(true);
            }
        } else {
            if (filterAttack == view) {
                getModel().sort(new FilterData(Arrays.asList(FilterData.FILTER_ATTACK)));
            } else if (filterDefence == view) {
                getModel().sort(new FilterData(Arrays.asList(FilterData.FILTER_DEFENCE)));
            } else if (filterSpeed == view) {
                getModel().sort(new FilterData(Arrays.asList(FilterData.FILTER_SPEED)));
            }
        }
    }

    @Override
    public void onRefresh() {
        scrollListener.resetState();
        getModel().reload();
    }

    @Override
    public void onDestroy() {
        onDestroyDisposables.clear();
        super.onDestroy();
    }

    @Override
    public void setProgressVisibility(boolean isVisible) {
        if (refreshSwipe != null && refreshSwipe.isRefreshing() != isVisible) {
            refreshSwipe.setRefreshing(isVisible);
        }
    }

    @Override
    public void setItems(List<PokemonFullDataSchema> list) {
        adapter.setItems(list);
    }

    @Override
    public void showError(Throwable error) {
        if (hasHost()) {
            getFragmentHost().showError(error);
        }
    }
}

File: app/src/main/java/com/mdgd/pokemon/ui/pokemons/PokemonsFragment.kt
package com.mdgd.pokemon.ui.pokemons

import android.os.Bundle
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.ImageButton
import androidx.core.content.ContextCompat
import androidx.lifecycle.ViewModelProvider
import androidx.lifecycle.lifecycleScope
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.mdgd.mvi.fragments.HostedFragment
import com.mdgd.pokemon.PokemonsApp
import com.mdgd.pokemon.R
import com.mdgd.pokemon.models.filters.FilterData
import com.mdgd.pokemon.ui.pokemons.adapter.Pokemon
import com.mdgd.pokemon.ui.pokemons.adapter.PokemonsAdapter
import com.mdgd.pokemon.ui.pokemons.infra.EndlessScrollListener
import kotlinx.coroutines.launch

class PokemonsFragment : HostedFragment<
        PokemonsContract.View,
        PokemonsContract.ViewModel,
        PokemonsContract.Host>(),
    PokemonsContract.View, View.OnClickListener, SwipeRefreshLayout.OnRefreshListener {

    private val adapter = PokemonsAdapter()
    private var refreshSwipe: SwipeRefreshLayout? = null

    // maybe paging library?
    private val scrollListener: EndlessScrollListener = object : EndlessScrollListener() {
        override fun onLoadMore(page: Int, totalItemsCount: Int, view: RecyclerView?) {
            model?.loadPage(page)
        }
    }
    private var recyclerView: RecyclerView? = null
    private var filterAttack: ImageButton? = null
    private var filterDefence: ImageButton? = null
    private var filterSpeed: ImageButton? = null
    private var refresh: View? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        lifecycleScope.launch {
            adapter.getEventsObservable().collect {
                model?.onItemClicked(it.model)
            }
        }
    }

    override fun createModel(): PokemonsContract.ViewModel {
        return ViewModelProvider(
            this, PokemonsViewModelFactory(PokemonsApp.instance?.appComponent!!)
        )[PokemonsViewModel::class.java]
    }

    override fun onCreateView(
        inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?
    ): View? {
        return inflater.inflate(R.layout.fragment_pokemons, container, false)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        recyclerView = view.findViewById(R.id.pokemons_recycler)
        recyclerView?.layoutManager = LinearLayoutManager(activity)
        recyclerView?.addOnScrollListener(scrollListener)
        recyclerView?.adapter = adapter

        refresh = view.findViewById(R.id.pokemons_refresh)
        refreshSwipe = view.findViewById(R.id.pokemons_swipe_refresh)
        filterAttack = view.findViewById(R.id.pokemons_filter_attack)
        filterDefence = view.findViewById(R.id.pokemons_filter_defence)
        filterSpeed = view.findViewById(R.id.pokemons_filter_movement)

        refresh?.setOnClickListener(this)
        refreshSwipe?.setOnRefreshListener(this)
        filterAttack?.setOnClickListener(this)
        filterDefence?.setOnClickListener(this)
        filterSpeed?.setOnClickListener(this)
    }

    override fun proceedToNextScreen(pokemonId: Long?) {
        fragmentHost?.proceedToPokemonScreen(pokemonId)
    }

    override fun updateFilterButtons(activateFilter: Boolean, filter: String) {
        val view = when (filter) {
            FilterData.FILTER_ATTACK -> filterAttack
            FilterData.FILTER_DEFENCE -> filterDefence
            FilterData.FILTER_SPEED -> filterSpeed
            else -> null
        }

        if (activateFilter) {
            view?.setColorFilter(ContextCompat.getColor(requireContext(), R.color.filter_active))
        } else {
            view?.setColorFilter(ContextCompat.getColor(requireContext(), R.color.filter_inactive))
        }
    }

    override fun onClick(view: View) {
        if (view === refresh) {
            onRefresh()
        } else {
            when {
                filterAttack === view -> model?.sort(FilterData.FILTER_ATTACK)
                filterDefence === view -> model?.sort(FilterData.FILTER_DEFENCE)
                filterSpeed === view -> model?.sort(FilterData.FILTER_SPEED)
            }
        }
    }

    override fun onRefresh() {
        scrollListener.resetState()
        model?.reload()
    }

    override fun setProgressVisibility(isVisible: Boolean) {
        if (refreshSwipe?.isRefreshing != isVisible) {
            refreshSwipe?.isRefreshing = isVisible
        }
    }

    override fun setItems(list: List<Pokemon>) {
        adapter.submitList(list)
    }

    override fun scrollToStart() {
        recyclerView?.scrollToPosition(0)
    }

    override fun showError(error: Throwable?) {
        fragmentHost?.showError(error)
    }
}
